File: ServerSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Scan_c_PatternDetectionPlayerDir/PatternDetectionPlayer.py
import os, sys
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class PatternDetectionPlayer(SuperPlayer):
    def __init__(self):
        super().__init__()
        self.my_name = None  # 必ずNoneで初期化
        self.patterns_detected = False  # パターン検出結果を保持
        self.binary_matrix_2Dlist = None
        self.png_file_path = None
        self.mode_charNumInfo_checksum_bitlist = None
        self.version = None
        self.grid_size = None
        self.qr_code_map = None
        self.modified_qr_code_map = None

    def return_my_name(self):
        return "PatternDetectionPlayer"

    def main(self):
        """
        25x25のバイナリマトリックスからパターンを検出します。
        """
        """ 入力 """
        woP ＝ self.one_time_world_instance.trapezoidCorrectionPlayer
        self.binary_matrix_2Dlist = woP.binary_matrix_2Dlist
        self.png_file_path = woP.png_file_path
        self.mode_charNumInfo_checksum_bitlist = woP.mode_charNumInfo_checksum_bitlist
        self.version = woP.mode_charNumInfo_checksum_bitlist
        self.grid_size = woP.grid_size
        self.qr_code_map = woP.qr_code_map
        self.modified_qr_code_map = woP.modified_qr_code_map
        
        
        # TrapezoidCorrectionPlayerから25x25のマトリックスを取得
        binary_matrix = self.one_time_world_instance.trapezoidCorrectionPlayer.binary_matrix_2Dlist

        # 入力データの検証
        if binary_matrix is None or len(binary_matrix) != 25 or len(binary_matrix[0]) != 25:
            raise ValueError("バイナリマトリックスが正しく取得できませんでした。")

        # パターン検出
        position_detected = self.detect_position_patterns(binary_matrix)
        #alignment_detected = self.detect_alignment_pattern(binary_matrix) # 現在不使用中
        timing_detected = self.detect_timing_patterns(binary_matrix)
        dark_module_detected = self.detect_dark_module(binary_matrix)

        # すべてのパターンが検出された場合
        if position_detected and alignment_detected and timing_detected:
            self.patterns_detected = True
            print("すべてのパターンが正常に検出されました。")
        else:
            print("パターンの検出に失敗しました。")
            if not position_detected:
                print("位置検出パターンの検出に失敗しました。")
            #if not alignment_detected:
                #print("アライメントパターンの検出に失敗しました。") #現在不使用中
            if not timing_detected:
                print("タイミングパターンの検出に失敗しました。")
            if not dark_module_detected:
                print("ダークモジュールの検出に失敗しました。")
                
        """ 出力 """
        self.binary_matrix_2Dlist
        self.png_file_path
        self.mode_charNumInfo_checksum_bitlist
        self.version
        self.grid_size
        self.qr_code_map
        self.modified_qr_code_map

        # 自身を更新
        self.one_time_world_instance.patternDetectionPlayer = self

        return "Completed"

    def detect_position_patterns(self, matrix):
        """
        位置検出パターンを検出します。

        Args:
            matrix (list): 25x25のバイナリマトリックス。

        Returns:
            bool: パターンが正しく検出された場合はTrue、そうでない場合はFalse。
        """
        # 位置検出パターンの座標
        positions = [
            (0, 0),               # 左上
            (0, len(matrix)-7),   # 右上
            (len(matrix)-7, 0),   # 左下
        ]

        for (row_start, col_start) in positions:
            if not self.check_position_detection_pattern(matrix, row_start, col_start):
                return False
        return True

    def check_position_detection_pattern(self, matrix, row_start, col_start):
        """
        単一の位置検出パターンを検証します。

        Args:
            matrix (list): バイナリマトリックス。
            row_start (int): パターンの開始行。
            col_start (int): パターンの開始列。

        Returns:
            bool: パターンが正しければTrue、そうでなければFalse。
        """
        # 位置検出パターンのサイズは7x7
        pattern_size = 7
        expected_pattern = [
            [1, 1, 1, 1, 1, 1, 1],
            [1, 0, 0, 0, 0, 0, 1],
            [1, 0, 1, 1, 1, 0, 1],
            [1, 0, 1, 1, 1, 0, 1],
            [1, 0, 1, 1, 1, 0, 1],
            [1, 0, 0, 0, 0, 0, 1],
            [1, 1, 1, 1, 1, 1, 1],
        ]

        for i in range(pattern_size):
            for j in range(pattern_size):
                if matrix[row_start + i][col_start + j] != expected_pattern[i][j]:
                    return False
        return True

    def detect_alignment_pattern(self, matrix):
        """
        アライメントパターンを検出します。

        Args:
            matrix (list): 25x25のバイナリマトリックス。

        Returns:
            bool: パターンが正しく検出された場合はTrue、そうでない場合はFalse。
        """
        # アライメントパターンの中心座標（モデル2 バージョン2の場合）
        center_row = 19
        center_col = 19

        # アライメントパターンのサイズは5x5
        pattern_size = 5
        row_start = center_row - 2
        col_start = center_col - 2

        expected_pattern = [
            [1, 1, 1, 1, 1],
            [1, 0, 0, 0, 1],
            [1, 0, 1, 0, 1],
            [1, 0, 0, 0, 1],
            [1, 1, 1, 1, 1],
        ]

        for i in range(pattern_size):
            for j in range(pattern_size):
                if matrix[row_start + i][col_start + j] != expected_pattern[i][j]:
                    return False
        return True

    def detect_timing_patterns(self, matrix):
        """
        タイミングパターンを検出します。

        Args:
            matrix (list): 25x25のバイナリマトリックス。

        Returns:
            bool: パターンが正しく検出された場合はTrue、そうでない場合はFalse。
        """
        size = len(matrix)

        # 行方向のタイミングパターンは (6, 8) から (6, size - 9) まで
        for col in range(8, size - 8):
            expected_bit = (col % 2)
            if matrix[6][col] != expected_bit:
                return False

        # 列方向のタイミングパターンは (8, 6) から (size - 9, 6) まで
        for row in range(8, size - 8):
            expected_bit = (row % 2)
            if matrix[row][6] != expected_bit:
                return False

        return True
        
    def detect_dark_module(self, matrix):
        """
        ダークモジュールを検出します。
    
        Args:
            matrix (list): 25x25のバイナリマトリックス。
    
        Returns:
            bool: ダークモジュールが正しく検出された場合はTrue、そうでない場合はFalse。
        """
        # ダークモジュールの位置は (8, 13) で常に黒（1）であるべき
        dark_module_position = (8, 13)
        
        # ダークモジュールが黒（1）であればTrue、そうでなければFalse
        return matrix[dark_module_position[0]][dark_module_position[1]] == 1

File: InstanceSide/BallPassSchedulePattern/Players_QrcodeReaderTypes/Scan_c_PatternDetectionPlayerDir/PatternDetectionPlayer.cs
using UdonSharp;
using UnityEngine;

public class PatternDetectionPlayer : SuperPlayer
{
    // Unityエディタでアタッチするオブジェクト
    public GameObject textureSource; // 画像を持つオブジェクト（マテリアルからテクスチャを取得）
    public RinaNumpy rinaNumpy; // RinaNumpyインスタンスをアタッチ

    public string pngFilePath; // 写真のパス
    public int[][] binaryMatrix2DList; // バイナリマトリックス (25x25)
    public bool patternsDetected; // パターン検出結果

    public override string ReturnMyName()
    {
        return "PatternDetectionPlayer";
    }

    public override string ExecuteMain()
    {
        // 入力データの取得
        Texture2D texture = GetTextureFromGameObject(textureSource);

        if (texture == null)
        {
            Debug.LogError("Texture could not be retrieved from the provided GameObject.");
            return "Error";
        }

        // マトリックスデータの仮の取得（具体的には外部プロセスで生成される想定）
        binaryMatrix2DList = GetBinaryMatrixFromTexture(texture);

        if (binaryMatrix2DList == null || binaryMatrix2DList.Length != 25 || binaryMatrix2DList[0].Length != 25)
        {
            Debug.LogError("Binary matrix is invalid or not properly formatted.");
            return "Error";
        }

        // パターン検出処理
        bool positionDetected = DetectPositionPatterns(binaryMatrix2DList);
        bool timingDetected = DetectTimingPatterns(binaryMatrix2DList);
        bool darkModuleDetected = DetectDarkModule(binaryMatrix2DList);

        // 結果の確認
        patternsDetected = positionDetected && timingDetected && darkModuleDetected;

        if (patternsDetected)
        {
            Debug.Log("All patterns successfully detected.");
        }
        else
        {
            if (!positionDetected)
                Debug.LogError("Failed to detect position patterns.");
            if (!timingDetected)
                Debug.LogError("Failed to detect timing patterns.");
            if (!darkModuleDetected)
                Debug.LogError("Failed to detect dark module.");
        }

        return "Completed";
    }

    private Texture2D GetTextureFromGameObject(GameObject obj)
    {
        if (obj == null)
            return null;

        Renderer renderer = obj.GetComponent<Renderer>();
        if (renderer != null && renderer.material != null)
        {
            return renderer.material.mainTexture as Texture2D;
        }
        return null;
    }

    private int[][] GetBinaryMatrixFromTexture(Texture2D texture)
    {
        // 仮の実装: テクスチャを解析してバイナリマトリックスを生成する
        int size = 25; // 固定サイズ
        int[][] matrix = new int[size][];
        for (int i = 0; i < size; i++)
        {
            matrix[i] = new int[size];
            for (int j = 0; j < size; j++)
            {
                // テクスチャのピクセルから値を取得する処理（適宜調整）
                Color pixelColor = texture.GetPixel(i, j);
                matrix[i][j] = pixelColor.grayscale > 0.5f ? 1 : 0; // 仮の閾値
            }
        }
        return matrix;
    }

    private bool DetectPositionPatterns(int[][] matrix)
    {
        // 位置検出パターンの座標
        int[][] positions = {
            new int[] {0, 0}, // 左上
            new int[] {0, matrix.Length - 7}, // 右上
            new int[] {matrix.Length - 7, 0} // 左下
        };

        foreach (int[] position in positions)
        {
            if (!CheckPositionDetectionPattern(matrix, position[0], position[1]))
            {
                return false;
            }
        }
        return true;
    }

    private bool CheckPositionDetectionPattern(int[][] matrix, int rowStart, int colStart)
    {
        int patternSize = 7;
        int[][] expectedPattern = {
            new int[] {1, 1, 1, 1, 1, 1, 1},
            new int[] {1, 0, 0, 0, 0, 0, 1},
            new int[] {1, 0, 1, 1, 1, 0, 1},
            new int[] {1, 0, 1, 1, 1, 0, 1},
            new int[] {1, 0, 1, 1, 1, 0, 1},
            new int[] {1, 0, 0, 0, 0, 0, 1},
            new int[] {1, 1, 1, 1, 1, 1, 1}
        };

        for (int i = 0; i < patternSize; i++)
        {
            // RinaNumpyを使って簡略化
            if (!rinaNumpy.CompareIntArrays(matrix[rowStart + i], expectedPattern[i], patternSize))
            {
                return false;
            }
        }
        return true;
    }

    private bool DetectTimingPatterns(int[][] matrix)
    {
        int size = matrix.Length;

        // 行方向のタイミングパターン
        for (int col = 8; col < size - 8; col++)
        {
            int expectedBit = col % 2;
            if (matrix[6][col] != expectedBit)
            {
                return false;
            }
        }

        // 列方向のタイミングパターン
        for (int row = 8; row < size - 8; row++)
        {
            int expectedBit = row % 2;
            if (matrix[row][6] != expectedBit)
            {
                return false;
            }
        }

        return true;
    }

    private bool DetectDarkModule(int[][] matrix)
    {
        // ダークモジュールの位置は (8, 13)
        return matrix[8][13] == 1;
    }
}
